core/src/main/java/org/eclipse/hono/util/HonoProtonHelper.java

```diff
@@ -144,10 +144,10 @@ public static void setDefaultCloseHandler(final ProtonSession session) {
      */
     public static boolean isLinkEstablished(final ProtonLink<?> link) {
         if (link instanceof ProtonSender) {
-            // check for non-null remote target address - or if it is null, for an anonymous sender link (with empty local target address)
+            // check for non-null remote target address - or if it is null, for an anonymous sender link (with null local target address) or an empty target address
             return link.getRemoteTarget() != null
                     && (link.getRemoteTarget().getAddress() != null
-                            || (link.getTarget() != null && link.getTarget().getAddress() != null && link.getTarget().getAddress().isEmpty()));
+                            || (link.getTarget() != null && Strings.isNullOrEmpty(link.getTarget().getAddress())));
```

Comment:
My understanding of the Anonymous Relay spec is that the corresponding target address is `null` and not the empty string. If we agree on that, then I would rather only check for a `null` target address here and instead fix the places in Hono where we establish an anonymous sender link with the dispatch router accordingly. AFAIK we do not explicitly state anywhere in Hono's API that we accept an empty string instead of `null` as the target address to indicate an anonymous sender, do we? In fact, IMHO we should use this issue to amend the AMQP User Guide with appropriate wording making clear that we expect a client to use `null` as the target address instead of an emtpy string ...

Reply:
> My understanding of the Anonymous Relay spec is that the corresponding target address is `null` and not the empty string. If we agree on that, then I would rather only check for a `null` target address here

So you mean not supporting the empty target address any more, correct?
We should reject this case in `createSender()` then already ("unsupported address"), IMHO.

> instead fix the places in Hono where we establish an anonymous sender link with the dispatch router accordingly

There are no such places any more AFAIK. The `DelegatedCommandSenderImpl` used this but was changed to use an explicit target address.

> In fact, IMHO we should use this issue to amend the AMQP User Guide [...]

Yes, I'll update that.

Reply:
> So you mean not supporting the empty target address any more, correct?

Yes, we should only consider the `null` target address as the one targeting the anonymous relay.

> We should reject this case in createSender() then already ("unsupported address"), IMHO.

Not sure if I understand correctly. Do you mean to not accept `""` as the target address at all? Again, I would argue that the empty string is a valid address and should therefore be supported (though no Hono component uses it currently).

> There are no such places any more AFAIK. The DelegatedCommandSenderImpl used this but was changed to use an explicit target address.

Even better :-)

Reply:
> Not sure if I understand correctly. Do you mean to not accept `""` as the target address at all? Again, I would argue that the empty string is a valid address and should therefore be supported (though no Hono component uses it currently).

Well, if we don't not allow `link.getTarget().getAddress()` to be the empty string in the above code snippet, there will always be an exception in invoking `createSender()` with an empty string address. Which would mean we don't support it at all. 
That's why I deduced you wanted to remove support for it.

> Again, I would argue that the empty string is a valid address and should therefore be supported

Yes, it also would feel a bit strange rejecting a valid address (AMQP spec and Proton wise) in the generic HonoConnectionImpl class.

So, we keep it as suggested in this PR?

Reply:
No, but I think the *isLinkEstablished* check becomes much simpler for a sender:

```
return link.getRemoteTarget() != null;
```

or am I mistaken?

Reply:
@calohmn any additional thoughts on this?

Reply:
> No, but I think the isLinkEstablished check becomes much simpler for a sender:

Right, I've changed it to that.